### PR TITLE
HHH-13103 - Allow Hibernate Types to get access to the current configuration properties

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-13103

### CODE REVIEW

**Comment by vladmihalcea** [comment]
@sebersole The reason why I think it's worth to implement this feature is that we are currently configuring the Java and SQL Type Descriptors in the constructor of the `Type` instances. 

With this feature, we could allow the user to provide a specific constructor which takes the `TypeBootstrapContext` where we can pass the current `SessionFactory` properties or some other info we might need in future:

    public ArrayType(TypeBootstrapContext typeBootstrapContext) {
        super( VarcharTypeDescriptor.INSTANCE, ArrayTypeDescriptor.INSTANCE );
        this.settings = typeBootstrapContext.getConfigurationSettings();
    }

This way, we can get access to the current properties and the user can customize the custom `Type` object creation using the Hibernate properties.

One use case where I wanted this future is to customize the Jackson `ObjectMapper` used by the JSON Types. Currently, I can only do that via loading the `hibernate.properties` file externally, but not via the SessionFactory properties that might be passed by Spring or Spring Boot. This is [the original issue](https://github.com/vladmihalcea/hibernate-types/issues/35) that made me realize we could benefit from this new feature.

Let me know what you think.

**Comment by cberg-zalando** [comment]
@vladmihalcea I guess you will have to resolve the merge conflict before this moves any forward.

**Comment by vladmihalcea** [comment]
@cberg-zalando I no longer work for the Hibernate project, hence, I don't know what is the current status or in which branch contributions like this one have to be applied. If this one is not integrated into 5.4, then I'll provide a new PR for the 6.x branch once it becomes the main development branch.

**Comment by vladmihalcea** [comment]
@dreab8 I rebased this PR. From the upvotes this PR got, I think it's very useful and worth integrating.

Do you think it could be integrated in the next release? It's a small change and shouldn't be risky at all.

File: hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java [hunk 1]
```diff
@@ -88,7 +92,24 @@ public Type byClass(Class clazz, Properties parameters) {
...
+
+			Constructor<Type> bootstrapContextAwareTypeConstructor = ReflectHelper.getConstructor(
+					typeClass,
+					TypeBootstrapContext.class
+			);
+			if ( bootstrapContextAwareTypeConstructor != null ) {
+				ConfigurationService configurationService = typeConfiguration.getServiceRegistry().getService(
+						ConfigurationService.class );
+				Map<String, Object> configurationSettings = configurationService.getSettings();
+				type = bootstrapContextAwareTypeConstructor.newInstance( new TypeBootstrapContext(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java]: Not a fan of creating a new `TypeBootstrapContext` every time.  I'd pass it the `TypeConfiguration` and be done with it, or have `TypeConfiguration` implement `TypeBootstrapContext` (after making it an interface - which it ought to be anyway imo)

File: hibernate-core/src/test/java/org/hibernate/test/type/contributor/ArrayType.java [hunk 2]
```diff
@@ -33,4 +43,7 @@ public String getName() {
         return "comma-separated-array";
     }
 
+    public Map<String, Object> getSettings() {
```
Comment (sebersole) [inline hibernate-core/src/test/java/org/hibernate/test/type/contributor/ArrayType.java]: Not a fan of exposing these.  The whole point of the issue/PR is to allow Types to be settings-aware, not make the consumer of a Type settings-aware.  You expose it for a test, but I can guarantee with a fairly high-degree of certainty that someone will try to use this.  And at that consumer needs to protect for null every time (here for example with the static ref or non-parameterized ctor)

Comment (sebersole) [reply hibernate-core/src/test/java/org/hibernate/test/type/contributor/ArrayType.java]: Sorry, just realized this is a ArrayType from the tests, not the "main" one.

File: hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java (hunk 1 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java]: Or better yet, have TypeFactory be the TypeBootstrapContext

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java]: I can work on this change

Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java]: Thanks.

**Comment by sebersole** [comment]
Applied upstream with discussed change

**Pull request closed without merging**